{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "234ec8a2-5062-413d-9fec-44bb6c0d6bb1",
   "metadata": {},
   "source": [
    "This notebook will show how to map data using the MapMyCells data products released with the HMBA consensus Basal Ganglia taxonomy.\n",
    "\n",
    "We are going to start by installing the `cell_type_mapper` (actual backend for MapMyCells) and `abc_atlas_access` (convenient download of other data provided by the AllenInstitute) libraries. You may need to restart your kernel after running these cells so that the notebook will acknoweldge the installed libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9fe0e2e5-bdfc-41b0-a60f-e2936ee5ad2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2\n",
      "  Cloning https://github.com/alleninstitute/cell_type_mapper.git (to revision rc/v1.5.2) to /private/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/pip-install-8c53nk7s/cell-type-mapper_0bad1e0b8b984773bf66c09ace614cdc\n",
      "  Running command git clone --filter=blob:none --quiet https://github.com/alleninstitute/cell_type_mapper.git /private/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/pip-install-8c53nk7s/cell-type-mapper_0bad1e0b8b984773bf66c09ace614cdc\n",
      "  Running command git checkout -b rc/v1.5.2 --track origin/rc/v1.5.2\n",
      "  Switched to a new branch 'rc/v1.5.2'\n",
      "  branch 'rc/v1.5.2' set up to track 'origin/rc/v1.5.2'.\n",
      "  Resolved https://github.com/alleninstitute/cell_type_mapper.git to commit 2aefd39caf5799555c4bcad6ba437ca9d93d4bfd\n",
      "  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25hCollecting anndata>=0.11.0 (from cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached anndata-0.11.4-py3-none-any.whl.metadata (9.3 kB)\n",
      "Collecting argschema (from cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached argschema-3.0.4-py2.py3-none-any.whl\n",
      "Collecting matplotlib (from cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached matplotlib-3.10.1-cp312-cp312-macosx_11_0_arm64.whl.metadata (11 kB)\n",
      "Collecting marshmallow==3.12.2 (from cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached marshmallow-3.12.2-py2.py3-none-any.whl.metadata (7.8 kB)\n",
      "Collecting numpy (from cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached numpy-2.2.5-cp312-cp312-macosx_11_0_arm64.whl.metadata (62 kB)\n",
      "Collecting pytest (from cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached pytest-8.3.5-py3-none-any.whl.metadata (7.6 kB)\n",
      "Collecting pytest-xdist (from cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached pytest_xdist-3.6.1-py3-none-any.whl.metadata (4.3 kB)\n",
      "Collecting scipy<1.15.0 (from cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached scipy-1.14.1-cp312-cp312-macosx_12_0_arm64.whl.metadata (60 kB)\n",
      "Collecting array-api-compat!=1.5,>1.4 (from anndata>=0.11.0->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached array_api_compat-1.11.2-py3-none-any.whl.metadata (1.9 kB)\n",
      "Collecting h5py>=3.7 (from anndata>=0.11.0->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached h5py-3.13.0-cp312-cp312-macosx_11_0_arm64.whl.metadata (2.5 kB)\n",
      "Collecting natsort (from anndata>=0.11.0->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached natsort-8.4.0-py3-none-any.whl.metadata (21 kB)\n",
      "Requirement already satisfied: packaging>=24.2 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from anndata>=0.11.0->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2) (25.0)\n",
      "Collecting pandas!=2.1.0rc0,!=2.1.2,>=1.4 (from anndata>=0.11.0->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached pandas-2.2.3-cp312-cp312-macosx_11_0_arm64.whl.metadata (89 kB)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from pandas!=2.1.0rc0,!=2.1.2,>=1.4->anndata>=0.11.0->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2) (2.9.0.post0)\n",
      "Collecting pytz>=2020.1 (from pandas!=2.1.0rc0,!=2.1.2,>=1.4->anndata>=0.11.0->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached pytz-2025.2-py2.py3-none-any.whl.metadata (22 kB)\n",
      "Collecting tzdata>=2022.7 (from pandas!=2.1.0rc0,!=2.1.2,>=1.4->anndata>=0.11.0->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached tzdata-2025.2-py2.py3-none-any.whl.metadata (1.4 kB)\n",
      "Requirement already satisfied: six>=1.5 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from python-dateutil>=2.8.2->pandas!=2.1.0rc0,!=2.1.2,>=1.4->anndata>=0.11.0->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2) (1.17.0)\n",
      "Requirement already satisfied: pyyaml in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from argschema->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2) (6.0.2)\n",
      "Collecting contourpy>=1.0.1 (from matplotlib->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached contourpy-1.3.2-cp312-cp312-macosx_11_0_arm64.whl.metadata (5.5 kB)\n",
      "Collecting cycler>=0.10 (from matplotlib->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached cycler-0.12.1-py3-none-any.whl.metadata (3.8 kB)\n",
      "Collecting fonttools>=4.22.0 (from matplotlib->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached fonttools-4.57.0-cp312-cp312-macosx_10_13_universal2.whl.metadata (102 kB)\n",
      "Collecting kiwisolver>=1.3.1 (from matplotlib->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached kiwisolver-1.4.8-cp312-cp312-macosx_11_0_arm64.whl.metadata (6.2 kB)\n",
      "Collecting pillow>=8 (from matplotlib->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached pillow-11.2.1-cp312-cp312-macosx_11_0_arm64.whl.metadata (8.9 kB)\n",
      "Collecting pyparsing>=2.3.1 (from matplotlib->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached pyparsing-3.2.3-py3-none-any.whl.metadata (5.0 kB)\n",
      "Collecting iniconfig (from pytest->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached iniconfig-2.1.0-py3-none-any.whl.metadata (2.7 kB)\n",
      "Collecting pluggy<2,>=1.5 (from pytest->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached pluggy-1.5.0-py3-none-any.whl.metadata (4.8 kB)\n",
      "Collecting execnet>=2.1 (from pytest-xdist->cell_type_mapper@ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2)\n",
      "  Using cached execnet-2.1.1-py3-none-any.whl.metadata (2.9 kB)\n",
      "Using cached marshmallow-3.12.2-py2.py3-none-any.whl (47 kB)\n",
      "Using cached scipy-1.14.1-cp312-cp312-macosx_12_0_arm64.whl (29.9 MB)\n",
      "Using cached numpy-2.2.5-cp312-cp312-macosx_11_0_arm64.whl (14.2 MB)\n",
      "Using cached anndata-0.11.4-py3-none-any.whl (144 kB)\n",
      "Using cached array_api_compat-1.11.2-py3-none-any.whl (53 kB)\n",
      "Using cached h5py-3.13.0-cp312-cp312-macosx_11_0_arm64.whl (2.9 MB)\n",
      "Using cached pandas-2.2.3-cp312-cp312-macosx_11_0_arm64.whl (11.4 MB)\n",
      "Using cached pytz-2025.2-py2.py3-none-any.whl (509 kB)\n",
      "Using cached tzdata-2025.2-py2.py3-none-any.whl (347 kB)\n",
      "Using cached matplotlib-3.10.1-cp312-cp312-macosx_11_0_arm64.whl (8.0 MB)\n",
      "Using cached contourpy-1.3.2-cp312-cp312-macosx_11_0_arm64.whl (255 kB)\n",
      "Using cached cycler-0.12.1-py3-none-any.whl (8.3 kB)\n",
      "Using cached fonttools-4.57.0-cp312-cp312-macosx_10_13_universal2.whl (2.8 MB)\n",
      "Using cached kiwisolver-1.4.8-cp312-cp312-macosx_11_0_arm64.whl (65 kB)\n",
      "Using cached pillow-11.2.1-cp312-cp312-macosx_11_0_arm64.whl (3.0 MB)\n",
      "Using cached pyparsing-3.2.3-py3-none-any.whl (111 kB)\n",
      "Using cached natsort-8.4.0-py3-none-any.whl (38 kB)\n",
      "Using cached pytest-8.3.5-py3-none-any.whl (343 kB)\n",
      "Using cached pluggy-1.5.0-py3-none-any.whl (20 kB)\n",
      "Using cached iniconfig-2.1.0-py3-none-any.whl (6.0 kB)\n",
      "Using cached pytest_xdist-3.6.1-py3-none-any.whl (46 kB)\n",
      "Using cached execnet-2.1.1-py3-none-any.whl (40 kB)\n",
      "Building wheels for collected packages: cell_type_mapper\n",
      "  Building wheel for cell_type_mapper (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for cell_type_mapper: filename=cell_type_mapper-1.5.2-py3-none-any.whl size=1267585 sha256=c64b8f0d1a232526b33ffc1153f30c67df0c81f1987fc6326ff9069a99a1f511\n",
      "  Stored in directory: /private/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/pip-ephem-wheel-cache-w2bychy5/wheels/39/4b/ba/4a4748790aa5c8131acd28a37fb39150691656e4b2ca693527\n",
      "Successfully built cell_type_mapper\n",
      "Installing collected packages: pytz, tzdata, pyparsing, pluggy, pillow, numpy, natsort, marshmallow, kiwisolver, iniconfig, fonttools, execnet, cycler, array-api-compat, scipy, pytest, pandas, h5py, contourpy, argschema, pytest-xdist, matplotlib, anndata, cell_type_mapper\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m24/24\u001b[0m [cell_type_mapper][cell_type_mapper]\n",
      "\u001b[1A\u001b[2KSuccessfully installed anndata-0.11.4 argschema-3.0.4 array-api-compat-1.11.2 cell_type_mapper-1.5.2 contourpy-1.3.2 cycler-0.12.1 execnet-2.1.1 fonttools-4.57.0 h5py-3.13.0 iniconfig-2.1.0 kiwisolver-1.4.8 marshmallow-3.12.2 matplotlib-3.10.1 natsort-8.4.0 numpy-2.2.5 pandas-2.2.3 pillow-11.2.1 pluggy-1.5.0 pyparsing-3.2.3 pytest-8.3.5 pytest-xdist-3.6.1 pytz-2025.2 scipy-1.14.1 tzdata-2025.2\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install \"cell_type_mapper @ git+https://github.com/alleninstitute/cell_type_mapper.git@rc/v1.5.2\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "044dbaa2-eea7-4f71-bc36-6d33f47026bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git (from abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git)\n",
      "  Cloning https://github.com/alleninstitute/abc_atlas_access.git to /private/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/pip-install-i48hugce/abc-atlas-access_ddbb425a58264c86b248be905825cb7b\n",
      "  Running command git clone --filter=blob:none --quiet https://github.com/alleninstitute/abc_atlas_access.git /private/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/pip-install-i48hugce/abc-atlas-access_ddbb425a58264c86b248be905825cb7b\n",
      "  Resolved https://github.com/alleninstitute/abc_atlas_access.git to commit a82a6770c99ad166105c3e6fccd47f31ee69b44c\n",
      "  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: anndata in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (0.11.4)\n",
      "Collecting boto3 (from abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git)\n",
      "  Downloading boto3-1.38.11-py3-none-any.whl.metadata (6.6 kB)\n",
      "Requirement already satisfied: numpy in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (2.2.5)\n",
      "Requirement already satisfied: pandas in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (2.2.3)\n",
      "Collecting pydantic (from abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git)\n",
      "  Using cached pydantic-2.11.4-py3-none-any.whl.metadata (66 kB)\n",
      "Collecting tqdm (from abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git)\n",
      "  Using cached tqdm-4.67.1-py3-none-any.whl.metadata (57 kB)\n",
      "Requirement already satisfied: matplotlib in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (3.10.1)\n",
      "Requirement already satisfied: scipy in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (1.14.1)\n",
      "Collecting simpleitk (from abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git)\n",
      "  Downloading simpleitk-2.5.0-cp311-abi3-macosx_11_0_arm64.whl.metadata (7.2 kB)\n",
      "Requirement already satisfied: array-api-compat!=1.5,>1.4 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from anndata->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (1.11.2)\n",
      "Requirement already satisfied: h5py>=3.7 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from anndata->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (3.13.0)\n",
      "Requirement already satisfied: natsort in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from anndata->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (8.4.0)\n",
      "Requirement already satisfied: packaging>=24.2 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from anndata->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (25.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from pandas->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from pandas->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (2025.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from pandas->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (2025.2)\n",
      "Requirement already satisfied: six>=1.5 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from python-dateutil>=2.8.2->pandas->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (1.17.0)\n",
      "Collecting botocore<1.39.0,>=1.38.11 (from boto3->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git)\n",
      "  Downloading botocore-1.38.11-py3-none-any.whl.metadata (5.7 kB)\n",
      "Collecting jmespath<2.0.0,>=0.7.1 (from boto3->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git)\n",
      "  Using cached jmespath-1.0.1-py3-none-any.whl.metadata (7.6 kB)\n",
      "Collecting s3transfer<0.13.0,>=0.12.0 (from boto3->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git)\n",
      "  Using cached s3transfer-0.12.0-py3-none-any.whl.metadata (1.7 kB)\n",
      "Requirement already satisfied: urllib3!=2.2.0,<3,>=1.25.4 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from botocore<1.39.0,>=1.38.11->boto3->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (2.4.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from matplotlib->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (1.3.2)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from matplotlib->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from matplotlib->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (4.57.0)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from matplotlib->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (1.4.8)\n",
      "Requirement already satisfied: pillow>=8 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from matplotlib->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (11.2.1)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from matplotlib->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (3.2.3)\n",
      "Collecting annotated-types>=0.6.0 (from pydantic->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git)\n",
      "  Using cached annotated_types-0.7.0-py3-none-any.whl.metadata (15 kB)\n",
      "Collecting pydantic-core==2.33.2 (from pydantic->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git)\n",
      "  Using cached pydantic_core-2.33.2-cp312-cp312-macosx_11_0_arm64.whl.metadata (6.8 kB)\n",
      "Requirement already satisfied: typing-extensions>=4.12.2 in /Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages (from pydantic->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git) (4.13.2)\n",
      "Collecting typing-inspection>=0.4.0 (from pydantic->abc_atlas_access@ git+https://github.com/alleninstitute/abc_atlas_access.git->abc_atlas_access[notebooks]@ git+https://github.com/alleninstitute/abc_atlas_access.git)\n",
      "  Using cached typing_inspection-0.4.0-py3-none-any.whl.metadata (2.6 kB)\n",
      "Downloading boto3-1.38.11-py3-none-any.whl (139 kB)\n",
      "Downloading botocore-1.38.11-py3-none-any.whl (13.5 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m13.5/13.5 MB\u001b[0m \u001b[31m23.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hUsing cached jmespath-1.0.1-py3-none-any.whl (20 kB)\n",
      "Using cached s3transfer-0.12.0-py3-none-any.whl (84 kB)\n",
      "Using cached pydantic-2.11.4-py3-none-any.whl (443 kB)\n",
      "Using cached pydantic_core-2.33.2-cp312-cp312-macosx_11_0_arm64.whl (1.8 MB)\n",
      "Using cached annotated_types-0.7.0-py3-none-any.whl (13 kB)\n",
      "Using cached typing_inspection-0.4.0-py3-none-any.whl (14 kB)\n",
      "Downloading simpleitk-2.5.0-cp311-abi3-macosx_11_0_arm64.whl (38.5 MB)\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m38.5/38.5 MB\u001b[0m \u001b[31m11.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hUsing cached tqdm-4.67.1-py3-none-any.whl (78 kB)\n",
      "Building wheels for collected packages: abc_atlas_access\n",
      "  Building wheel for abc_atlas_access (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for abc_atlas_access: filename=abc_atlas_access-0.6.0-py3-none-any.whl size=21653 sha256=50a9d13cfa58e3ca82db8f53725053f420b444ceb17885c3a53c24cc6ea6ec3b\n",
      "  Stored in directory: /private/var/folders/8b/hnw5vq8s20jbpz51wdhd11fr0000gp/T/pip-ephem-wheel-cache-tfbr4h4i/wheels/0f/79/3e/7355e6338e46308d9510a8e2f699a9b8d69553bd796698f725\n",
      "Successfully built abc_atlas_access\n",
      "Installing collected packages: simpleitk, typing-inspection, tqdm, pydantic-core, jmespath, annotated-types, pydantic, botocore, s3transfer, boto3, abc_atlas_access\n",
      "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m11/11\u001b[0m [abc_atlas_access][boto3]re]\n",
      "\u001b[1A\u001b[2KSuccessfully installed abc_atlas_access-0.6.0 annotated-types-0.7.0 boto3-1.38.11 botocore-1.38.11 jmespath-1.0.1 pydantic-2.11.4 pydantic-core-2.33.2 s3transfer-0.12.0 simpleitk-2.5.0 tqdm-4.67.1 typing-inspection-0.4.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install \"abc_atlas_access[notebooks] @ git+https://github.com/alleninstitute/abc_atlas_access.git\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e262f008-c7fa-4de3-bd90-8bed9d9f3ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "import cell_type_mapper\n",
    "from abc_atlas_access.abc_atlas_cache.abc_project_cache import AbcProjectCache"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1c5ec60-cd95-47c5-bb15-6ee763a2fd18",
   "metadata": {},
   "source": [
    "To run MapMyCells, you need two supporting data files: [an HDF5 file](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/input_data_files/precomputed_stats_file.md) which defines your taxonomy and the average gene expression profiles of the taxons there in, and a [lookup table of marker genes](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/input_data_files/marker_gene_lookup.md) for your taxonomy. For the sake of this exercise, let's download those two files for the human basal ganglia data. Refer back [to the main page](https://github.com/AllenInstitute/HMBA_BasalGanglia_Consensus_Taxonomy/blob/main/index.md#cell-type-mapping-with-mapmycells) for the locations of the relevant files for all three species (human, marmoset, and macaque)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "728a211a-527f-4d47-b94f-cdf414fded84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-05-08 10:07:31--  https://released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com/HMBA/BasalGanglia/BICAN_05072025_pre-print_release/MapMyCells/Human.precomputed_stats.20250507.h5\n",
      "Resolving released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com (released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com)... 3.5.85.43, 52.218.232.209, 52.218.153.41, ...\n",
      "Connecting to released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com (released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com)|3.5.85.43|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 755535936 (721M) [binary/octet-stream]\n",
      "Saving to: ‘data/Human.precomputed_stats.20250507.h5’\n",
      "\n",
      "data/Human.precompu 100%[===================>] 720.54M  12.5MB/s    in 60s     \n",
      "\n",
      "2025-05-08 10:08:31 (12.0 MB/s) - ‘data/Human.precomputed_stats.20250507.h5’ saved [755535936/755535936]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! wget \"https://released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com/HMBA/BasalGanglia/BICAN_05072025_pre-print_release/MapMyCells/Human.precomputed_stats.20250507.h5\" -O \"data/Human.precomputed_stats.20250507.h5\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "da3c9616-435e-4ed7-a100-56f360979857",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-05-08 10:38:10--  https://released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com/HMBA/BasalGanglia/BICAN_05072025_pre-print_release/MapMyCells/Human.query_markers.20250507.json\n",
      "Resolving released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com (released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com)... 52.218.232.161, 52.92.133.26, 3.5.82.215, ...\n",
      "Connecting to released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com (released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com)|52.218.232.161|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 837105 (817K) [application/json]\n",
      "Saving to: ‘data/Human.query_markers.20250507.json’\n",
      "\n",
      "data/Human.query_ma 100%[===================>] 817.49K  --.-KB/s    in 0.09s   \n",
      "\n",
      "2025-05-08 10:38:11 (8.79 MB/s) - ‘data/Human.query_markers.20250507.json’ saved [837105/837105]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! wget \"https://released-taxonomies-802451596237-us-west-2.s3.us-west-2.amazonaws.com/HMBA/BasalGanglia/BICAN_05072025_pre-print_release/MapMyCells/Human.query_markers.20250507.json\" -O \"data/Human.query_markers.20250507.json\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "851d343f-40f7-45ae-9438-cad35be99f21",
   "metadata": {},
   "source": [
    "You also need some unlabeled ata to map. We will use [the abc_atlas_access tool](https://alleninstitute.github.io/abc_atlas_access/intro.html) to download some Whole Human Brain data that we can map onto the Basal Ganglia taxonomy. **Note:** we are going to download the non-neuronal data only, not because it is interesting, but because it is small (5 GB as opposed to 32 GB for neuronal data)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b403ca0a-bc0e-4b8b-aef9-67f9245de3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "abc_cache = AbcProjectCache.from_cache_dir('abc_cache')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5b7953b2-0be8-4e44-85f1-af845b7d1adb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['WHB-10Xv3-Neurons/log2',\n",
       " 'WHB-10Xv3-Neurons/raw',\n",
       " 'WHB-10Xv3-Nonneurons/log2',\n",
       " 'WHB-10Xv3-Nonneurons/raw']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abc_cache.list_data_files(directory='WHB-10Xv3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "abc659b9-a12c-478a-be7b-06eaf89920a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WHB-10Xv3-Nonneurons-raw.h5ad: 100%|██████| 4.75G/4.75G [06:57<00:00, 11.4MMB/s]\n"
     ]
    }
   ],
   "source": [
    "unmapped_human_data_path = abc_cache.get_data_path(\n",
    "    directory='WHB-10Xv3',\n",
    "    file_name='WHB-10Xv3-Nonneurons/raw'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eacd0613-a698-4c7f-8224-ba15856c009b",
   "metadata": {},
   "source": [
    "Now we will actually [perform the mapping](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/mapping_cells.md)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "48b3f5bf-c6da-4233-82c3-fd96206035d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cell_type_mapper.cli.from_specified_markers import FromSpecifiedMarkersRunner"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "529c1e69-49e5-4c1a-aa25-6de2db06f413",
   "metadata": {},
   "outputs": [],
   "source": [
    "# paths to files where mapping output will be written\n",
    "json_dst_path = \"results/human_nonneurons_mapping.json\"\n",
    "csv_dst_path = \"results/human_nonneruons_mapping.csv\"\n",
    "\n",
    "# the lookup table of marker genes which we downloaded from S3\n",
    "query_marker_path = \"data/Human.query_markers.20250507.json\"\n",
    "\n",
    "# the human-specific precomputed stats file which we downloaded from S3\n",
    "precomputed_path = \"data/Human.precomputed_stats.20250507.h5\"\n",
    "\n",
    "config = {\n",
    "    \"query_path\": str(unmapped_human_data_path.resolve().absolute()),\n",
    "    \"extended_result_path\": json_dst_path,\n",
    "    \"csv_result_path\": csv_dst_path,\n",
    "    \"verbose_csv\": True,\n",
    "    \"query_markers\": {\n",
    "       \"serialized_lookup\": query_marker_path\n",
    "    },\n",
    "    \"precomputed_stats\": {\n",
    "        \"path\": precomputed_path\n",
    "    },\n",
    "    \"type_assignment\": {\n",
    "        \"n_processors\": 4,\n",
    "        \"normalization\": \"raw\",\n",
    "        \"bootstrap_factor\": 0.5,\n",
    "        \"bootstrap_iteration\": 100\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d680bfe0-78fe-45a4-8815-5e229fcc4275",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== Running Hierarchical Mapping 1.5.2 with config ===\n",
      "{\n",
      "  \"log_level\": \"ERROR\",\n",
      "  \"extended_result_path\": \"results/human_nonneurons_mapping.json\",\n",
      "  \"verbose_stdout\": true,\n",
      "  \"type_assignment\": {\n",
      "    \"bootstrap_factor\": 0.5,\n",
      "    \"log_level\": \"ERROR\",\n",
      "    \"bootstrap_factor_lookup\": null,\n",
      "    \"rng_seed\": 11235813,\n",
      "    \"chunk_size\": 10000,\n",
      "    \"bootstrap_iteration\": 100,\n",
      "    \"n_processors\": 4,\n",
      "    \"min_markers\": 10,\n",
      "    \"n_runners_up\": 5,\n",
      "    \"normalization\": \"raw\"\n",
      "  },\n",
      "  \"cloud_safe\": false,\n",
      "  \"hdf5_result_path\": null,\n",
      "  \"log_path\": null,\n",
      "  \"query_markers\": {\n",
      "    \"collapse_markers\": false,\n",
      "    \"log_level\": \"ERROR\",\n",
      "    \"serialized_lookup\": \"data/Human.query_markers.20250507.json\"\n",
      "  },\n",
      "  \"csv_result_path\": \"results/human_nonneruons_mapping.csv\",\n",
      "  \"extended_result_dir\": null,\n",
      "  \"query_gene_id_col\": null,\n",
      "  \"max_gb\": 100.0,\n",
      "  \"obsm_clobber\": false,\n",
      "  \"verbose_csv\": false,\n",
      "  \"drop_level\": null,\n",
      "  \"summary_metadata_path\": null,\n",
      "  \"map_to_ensembl\": false,\n",
      "  \"nodes_to_drop\": null,\n",
      "  \"precomputed_stats\": {\n",
      "    \"path\": \"data/Human.precomputed_stats.20250507.h5\",\n",
      "    \"log_level\": \"ERROR\"\n",
      "  },\n",
      "  \"tmp_dir\": null,\n",
      "  \"flatten\": false,\n",
      "  \"obsm_key\": null,\n",
      "  \"query_path\": \"/Users/scott.daniel/KnowledgeEngineering/HMBA_BasalGanglia_Consensus_Taxonomy/examples/abc_cache/expression_matrices/WHB-10Xv3/20240330/WHB-10Xv3-Nonneurons-raw.h5ad\"\n",
      "}\n",
      "ENV: is_torch_available: False\n",
      "ENV: is_cuda_available: False\n",
      "ENV: use_torch: False\n",
      "ENV: multiprocessing start method: spawn\n",
      "ENV: Python version: 3.12.9 | packaged by Anaconda, Inc. | (main, Feb  6 2025, 12:55:12) [Clang 14.0.6 ]\n",
      "ENV: anndata version: 0.11.4\n",
      "ENV: numpy version: 2.2.5\n",
      "BENCHMARK: spent 1.1048e-03 seconds validating config and copying data\n",
      "using ../Human.precomputed_stats.20250507.h5 for precomputed_stats\n",
      "reading taxonomy_tree from ../Human.precomputed_stats.20250507.h5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages/cell_type_mapper/cli/cli_log.py:73: UserWarning: numpy's internal parallelization is enabled. This could cause independent worker processes to compete for resources, degrading performance. We recommend setting the following environment variables to '1' to improve performance\n",
      "{\n",
      "  \"NUMEXPR_NUM_THREADS\": \"\",\n",
      "  \"MKL_NUM_THREADS\": \"\",\n",
      "  \"OMP_NUM_THREADS\": \"\"\n",
      "}\n",
      "  warnings.warn(msg)\n",
      "/Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages/cell_type_mapper/taxonomy/utils.py:253: UserWarning: This taxonomy has no mapping from leaf_node -> rows in the cell by gene matrix\n",
      "  warnings.warn(\"This taxonomy has no mapping from leaf_node -> rows \"\n",
      "/Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages/cell_type_mapper/cli/cli_log.py:73: UserWarning: 18 marker genes were not present in the query dataset. They have been ignored\n",
      "  warnings.warn(msg)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BENCHMARK: spent 5.9319e-02 seconds creating query marker cache\n",
      "Running CPU implementation of type assignment.\n",
      "40000 of 888263 cells in 1.75e+00 min; predict 3.70e+01 min of 3.88e+01 min left\n",
      "60000 of 888263 cells in 2.43e+00 min; predict 3.36e+01 min of 3.60e+01 min left\n",
      "80000 of 888263 cells in 3.01e+00 min; predict 3.04e+01 min of 3.34e+01 min left\n",
      "100000 of 888263 cells in 3.72e+00 min; predict 2.93e+01 min of 3.30e+01 min left\n",
      "120000 of 888263 cells in 4.42e+00 min; predict 2.83e+01 min of 3.27e+01 min left\n",
      "140000 of 888263 cells in 5.16e+00 min; predict 2.76e+01 min of 3.27e+01 min left\n",
      "160000 of 888263 cells in 5.79e+00 min; predict 2.63e+01 min of 3.21e+01 min left\n",
      "180000 of 888263 cells in 6.46e+00 min; predict 2.54e+01 min of 3.19e+01 min left\n",
      "200000 of 888263 cells in 7.29e+00 min; predict 2.51e+01 min of 3.24e+01 min left\n",
      "220000 of 888263 cells in 7.99e+00 min; predict 2.43e+01 min of 3.23e+01 min left\n",
      "240000 of 888263 cells in 8.64e+00 min; predict 2.33e+01 min of 3.20e+01 min left\n",
      "260000 of 888263 cells in 9.33e+00 min; predict 2.25e+01 min of 3.19e+01 min left\n",
      "280000 of 888263 cells in 9.95e+00 min; predict 2.16e+01 min of 3.16e+01 min left\n",
      "300000 of 888263 cells in 1.06e+01 min; predict 2.08e+01 min of 3.15e+01 min left\n",
      "320000 of 888263 cells in 1.14e+01 min; predict 2.03e+01 min of 3.17e+01 min left\n",
      "340000 of 888263 cells in 1.19e+01 min; predict 1.92e+01 min of 3.11e+01 min left\n",
      "360000 of 888263 cells in 1.25e+01 min; predict 1.84e+01 min of 3.09e+01 min left\n",
      "380000 of 888263 cells in 1.32e+01 min; predict 1.77e+01 min of 3.09e+01 min left\n",
      "400000 of 888263 cells in 1.39e+01 min; predict 1.69e+01 min of 3.08e+01 min left\n",
      "420000 of 888263 cells in 1.46e+01 min; predict 1.63e+01 min of 3.08e+01 min left\n",
      "440000 of 888263 cells in 1.53e+01 min; predict 1.56e+01 min of 3.09e+01 min left\n",
      "460000 of 888263 cells in 1.60e+01 min; predict 1.49e+01 min of 3.09e+01 min left\n",
      "480000 of 888263 cells in 1.66e+01 min; predict 1.41e+01 min of 3.08e+01 min left\n",
      "500000 of 888263 cells in 1.75e+01 min; predict 1.36e+01 min of 3.10e+01 min left\n",
      "520000 of 888263 cells in 1.81e+01 min; predict 1.28e+01 min of 3.10e+01 min left\n",
      "540000 of 888263 cells in 1.89e+01 min; predict 1.22e+01 min of 3.11e+01 min left\n",
      "560000 of 888263 cells in 1.97e+01 min; predict 1.16e+01 min of 3.13e+01 min left\n",
      "580000 of 888263 cells in 2.05e+01 min; predict 1.09e+01 min of 3.13e+01 min left\n",
      "600000 of 888263 cells in 2.14e+01 min; predict 1.03e+01 min of 3.17e+01 min left\n",
      "620000 of 888263 cells in 2.22e+01 min; predict 9.62e+00 min of 3.18e+01 min left\n",
      "640000 of 888263 cells in 2.30e+01 min; predict 8.93e+00 min of 3.20e+01 min left\n",
      "660000 of 888263 cells in 2.39e+01 min; predict 8.26e+00 min of 3.21e+01 min left\n",
      "680000 of 888263 cells in 2.45e+01 min; predict 7.50e+00 min of 3.20e+01 min left\n",
      "700000 of 888263 cells in 2.52e+01 min; predict 6.78e+00 min of 3.20e+01 min left\n",
      "720000 of 888263 cells in 2.58e+01 min; predict 6.03e+00 min of 3.19e+01 min left\n",
      "750000 of 888263 cells in 2.65e+01 min; predict 4.88e+00 min of 3.14e+01 min left\n",
      "770000 of 888263 cells in 2.73e+01 min; predict 4.19e+00 min of 3.15e+01 min left\n",
      "790000 of 888263 cells in 2.81e+01 min; predict 3.49e+00 min of 3.15e+01 min left\n",
      "810000 of 888263 cells in 2.88e+01 min; predict 2.78e+00 min of 3.15e+01 min left\n",
      "830000 of 888263 cells in 2.95e+01 min; predict 2.07e+00 min of 3.16e+01 min left\n",
      "850000 of 888263 cells in 3.02e+01 min; predict 1.36e+00 min of 3.16e+01 min left\n",
      "870000 of 888263 cells in 3.09e+01 min; predict 6.49e-01 min of 3.16e+01 min left\n",
      "BENCHMARK: spent 1.9071e+03 seconds assigning cell types\n",
      "Writing marker genes to output file\n",
      "MAPPING FROM SPECIFIED MARKERS RAN SUCCESSFULLY\n",
      "CLEANING UP\n"
     ]
    }
   ],
   "source": [
    "runner = FromSpecifiedMarkersRunner(\n",
    "    args=[],\n",
    "    input_data=config\n",
    ")\n",
    "runner.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "153381c4-5655-4d41-9b09-b54289427e86",
   "metadata": {},
   "source": [
    "# Output of mapping file\n",
    "\n",
    "The results of our mapping are now in two files: the csv file pointed to by `csv_dst_path` and the JSON file pointed to by `json_dst_path`. Dedicated documentation of the the contents of the mapping output [can be found here.](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/output.md)\n",
    "\n",
    "## CSV output file\n",
    "\n",
    "The CSV file is effectively just a dataframe. For every cell at every taxonomy level, you have its assigned cell type (both as a guaranteed unique \"label\" and a more human readable \"name\") along with quality metrics assessing the confidence in the mapping (see the detailed documentation above)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "75fde950-89eb-46e6-b9ce-84b8f98bd281",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_csv = pd.read_csv(csv_dst_path, comment='#')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "1cb1d69b-6372-494c-99cc-9358cf893ac4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_id</th>\n",
       "      <th>Neighborhood_label</th>\n",
       "      <th>Neighborhood_name</th>\n",
       "      <th>Neighborhood_bootstrapping_probability</th>\n",
       "      <th>Class_label</th>\n",
       "      <th>Class_name</th>\n",
       "      <th>Class_bootstrapping_probability</th>\n",
       "      <th>Subclass_label</th>\n",
       "      <th>Subclass_name</th>\n",
       "      <th>Subclass_bootstrapping_probability</th>\n",
       "      <th>Group_label</th>\n",
       "      <th>Group_name</th>\n",
       "      <th>Group_bootstrapping_probability</th>\n",
       "      <th>Cluster_label</th>\n",
       "      <th>Cluster_name</th>\n",
       "      <th>Cluster_alias</th>\n",
       "      <th>Cluster_bootstrapping_probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10X362_3:TCAGTGAGTATTGACC</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>OPC-Oligo</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_SUBCL_0022</td>\n",
       "      <td>Oligodendrocyte</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_GROUP_0025</td>\n",
       "      <td>Oligo OPALIN</td>\n",
       "      <td>0.69</td>\n",
       "      <td>CS20250428_CLUST_0227</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>Human-1</td>\n",
       "      <td>0.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10X362_5:TCCGTGTGTGAAAGTT</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>OPC-Oligo</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_SUBCL_0022</td>\n",
       "      <td>Oligodendrocyte</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_GROUP_0026</td>\n",
       "      <td>Oligo PLEKHG1</td>\n",
       "      <td>0.99</td>\n",
       "      <td>CS20250428_CLUST_0231</td>\n",
       "      <td>Human-15</td>\n",
       "      <td>Human-15</td>\n",
       "      <td>0.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10X362_5:CACGGGTAGAGCAGAA</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>OPC-Oligo</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_SUBCL_0022</td>\n",
       "      <td>Oligodendrocyte</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_GROUP_0026</td>\n",
       "      <td>Oligo PLEKHG1</td>\n",
       "      <td>0.58</td>\n",
       "      <td>CS20250428_CLUST_0231</td>\n",
       "      <td>Human-15</td>\n",
       "      <td>Human-15</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10X362_5:GATTCTTGTATGTCAC</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>OPC-Oligo</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_SUBCL_0022</td>\n",
       "      <td>Oligodendrocyte</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_GROUP_0026</td>\n",
       "      <td>Oligo PLEKHG1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_CLUST_0230</td>\n",
       "      <td>Human-12</td>\n",
       "      <td>Human-12</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10X362_6:AGGACTTGTATCCTTT</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_CLASS_0010</td>\n",
       "      <td>OPC-Oligo</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_SUBCL_0022</td>\n",
       "      <td>Oligodendrocyte</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_GROUP_0026</td>\n",
       "      <td>Oligo PLEKHG1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_CLUST_0231</td>\n",
       "      <td>Human-15</td>\n",
       "      <td>Human-15</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>888258</th>\n",
       "      <td>10X194_8:GAAATGAGTTCGGCTG</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_CLASS_0008</td>\n",
       "      <td>Immune</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_SUBCL_0018</td>\n",
       "      <td>Macrophage</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_GROUP_0001</td>\n",
       "      <td>BAM</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_CLUST_0157</td>\n",
       "      <td>Human-71</td>\n",
       "      <td>Human-71</td>\n",
       "      <td>0.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>888259</th>\n",
       "      <td>10X350_4:TTTACCATCGCACGAC</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_CLASS_0008</td>\n",
       "      <td>Immune</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_SUBCL_0018</td>\n",
       "      <td>Macrophage</td>\n",
       "      <td>0.95</td>\n",
       "      <td>CS20250428_GROUP_0001</td>\n",
       "      <td>BAM</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_CLUST_0160</td>\n",
       "      <td>Human-75</td>\n",
       "      <td>Human-75</td>\n",
       "      <td>0.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>888260</th>\n",
       "      <td>10X225_1:AGAAGCGTCCATATGG</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_CLASS_0008</td>\n",
       "      <td>Immune</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_SUBCL_0018</td>\n",
       "      <td>Macrophage</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_GROUP_0001</td>\n",
       "      <td>BAM</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_CLUST_0159</td>\n",
       "      <td>Human-74</td>\n",
       "      <td>Human-74</td>\n",
       "      <td>0.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>888261</th>\n",
       "      <td>10X221_5:TTGAACGCAGCCTTCT</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_CLASS_0008</td>\n",
       "      <td>Immune</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_SUBCL_0018</td>\n",
       "      <td>Macrophage</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_GROUP_0001</td>\n",
       "      <td>BAM</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_CLUST_0160</td>\n",
       "      <td>Human-75</td>\n",
       "      <td>Human-75</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>888262</th>\n",
       "      <td>10X385_3:CTACCCAGTGGCGCTT</td>\n",
       "      <td>CS20250428_NEIGH_0001</td>\n",
       "      <td>Nonneuron</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_CLASS_0008</td>\n",
       "      <td>Immune</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CS20250428_SUBCL_0018</td>\n",
       "      <td>Macrophage</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_GROUP_0001</td>\n",
       "      <td>BAM</td>\n",
       "      <td>1.00</td>\n",
       "      <td>CS20250428_CLUST_0158</td>\n",
       "      <td>Human-73</td>\n",
       "      <td>Human-73</td>\n",
       "      <td>0.40</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>888263 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          cell_id     Neighborhood_label Neighborhood_name  \\\n",
       "0       10X362_3:TCAGTGAGTATTGACC  CS20250428_NEIGH_0001         Nonneuron   \n",
       "1       10X362_5:TCCGTGTGTGAAAGTT  CS20250428_NEIGH_0001         Nonneuron   \n",
       "2       10X362_5:CACGGGTAGAGCAGAA  CS20250428_NEIGH_0001         Nonneuron   \n",
       "3       10X362_5:GATTCTTGTATGTCAC  CS20250428_NEIGH_0001         Nonneuron   \n",
       "4       10X362_6:AGGACTTGTATCCTTT  CS20250428_NEIGH_0001         Nonneuron   \n",
       "...                           ...                    ...               ...   \n",
       "888258  10X194_8:GAAATGAGTTCGGCTG  CS20250428_NEIGH_0001         Nonneuron   \n",
       "888259  10X350_4:TTTACCATCGCACGAC  CS20250428_NEIGH_0001         Nonneuron   \n",
       "888260  10X225_1:AGAAGCGTCCATATGG  CS20250428_NEIGH_0001         Nonneuron   \n",
       "888261  10X221_5:TTGAACGCAGCCTTCT  CS20250428_NEIGH_0001         Nonneuron   \n",
       "888262  10X385_3:CTACCCAGTGGCGCTT  CS20250428_NEIGH_0001         Nonneuron   \n",
       "\n",
       "        Neighborhood_bootstrapping_probability            Class_label  \\\n",
       "0                                          1.0  CS20250428_CLASS_0010   \n",
       "1                                          1.0  CS20250428_CLASS_0010   \n",
       "2                                          1.0  CS20250428_CLASS_0010   \n",
       "3                                          1.0  CS20250428_CLASS_0010   \n",
       "4                                          1.0  CS20250428_CLASS_0010   \n",
       "...                                        ...                    ...   \n",
       "888258                                     1.0  CS20250428_CLASS_0008   \n",
       "888259                                     1.0  CS20250428_CLASS_0008   \n",
       "888260                                     1.0  CS20250428_CLASS_0008   \n",
       "888261                                     1.0  CS20250428_CLASS_0008   \n",
       "888262                                     1.0  CS20250428_CLASS_0008   \n",
       "\n",
       "       Class_name  Class_bootstrapping_probability         Subclass_label  \\\n",
       "0       OPC-Oligo                              1.0  CS20250428_SUBCL_0022   \n",
       "1       OPC-Oligo                              1.0  CS20250428_SUBCL_0022   \n",
       "2       OPC-Oligo                              1.0  CS20250428_SUBCL_0022   \n",
       "3       OPC-Oligo                              1.0  CS20250428_SUBCL_0022   \n",
       "4       OPC-Oligo                              1.0  CS20250428_SUBCL_0022   \n",
       "...           ...                              ...                    ...   \n",
       "888258     Immune                              1.0  CS20250428_SUBCL_0018   \n",
       "888259     Immune                              1.0  CS20250428_SUBCL_0018   \n",
       "888260     Immune                              1.0  CS20250428_SUBCL_0018   \n",
       "888261     Immune                              1.0  CS20250428_SUBCL_0018   \n",
       "888262     Immune                              1.0  CS20250428_SUBCL_0018   \n",
       "\n",
       "          Subclass_name  Subclass_bootstrapping_probability  \\\n",
       "0       Oligodendrocyte                                1.00   \n",
       "1       Oligodendrocyte                                1.00   \n",
       "2       Oligodendrocyte                                1.00   \n",
       "3       Oligodendrocyte                                1.00   \n",
       "4       Oligodendrocyte                                1.00   \n",
       "...                 ...                                 ...   \n",
       "888258       Macrophage                                1.00   \n",
       "888259       Macrophage                                0.95   \n",
       "888260       Macrophage                                1.00   \n",
       "888261       Macrophage                                1.00   \n",
       "888262       Macrophage                                1.00   \n",
       "\n",
       "                  Group_label     Group_name  Group_bootstrapping_probability  \\\n",
       "0       CS20250428_GROUP_0025   Oligo OPALIN                             0.69   \n",
       "1       CS20250428_GROUP_0026  Oligo PLEKHG1                             0.99   \n",
       "2       CS20250428_GROUP_0026  Oligo PLEKHG1                             0.58   \n",
       "3       CS20250428_GROUP_0026  Oligo PLEKHG1                             1.00   \n",
       "4       CS20250428_GROUP_0026  Oligo PLEKHG1                             1.00   \n",
       "...                       ...            ...                              ...   \n",
       "888258  CS20250428_GROUP_0001            BAM                             1.00   \n",
       "888259  CS20250428_GROUP_0001            BAM                             1.00   \n",
       "888260  CS20250428_GROUP_0001            BAM                             1.00   \n",
       "888261  CS20250428_GROUP_0001            BAM                             1.00   \n",
       "888262  CS20250428_GROUP_0001            BAM                             1.00   \n",
       "\n",
       "                Cluster_label Cluster_name Cluster_alias  \\\n",
       "0       CS20250428_CLUST_0227      Human-1       Human-1   \n",
       "1       CS20250428_CLUST_0231     Human-15      Human-15   \n",
       "2       CS20250428_CLUST_0231     Human-15      Human-15   \n",
       "3       CS20250428_CLUST_0230     Human-12      Human-12   \n",
       "4       CS20250428_CLUST_0231     Human-15      Human-15   \n",
       "...                       ...          ...           ...   \n",
       "888258  CS20250428_CLUST_0157     Human-71      Human-71   \n",
       "888259  CS20250428_CLUST_0160     Human-75      Human-75   \n",
       "888260  CS20250428_CLUST_0159     Human-74      Human-74   \n",
       "888261  CS20250428_CLUST_0160     Human-75      Human-75   \n",
       "888262  CS20250428_CLUST_0158     Human-73      Human-73   \n",
       "\n",
       "        Cluster_bootstrapping_probability  \n",
       "0                                    0.55  \n",
       "1                                    0.99  \n",
       "2                                    1.00  \n",
       "3                                    1.00  \n",
       "4                                    1.00  \n",
       "...                                   ...  \n",
       "888258                               0.97  \n",
       "888259                               0.92  \n",
       "888260                               0.90  \n",
       "888261                               1.00  \n",
       "888262                               0.40  \n",
       "\n",
       "[888263 rows x 17 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0357fa7b-be1a-4906-b47d-f93031cc2be9",
   "metadata": {},
   "source": [
    "## JSON output file\n",
    "\n",
    "The JSON output contains everything in the CSV file, along with helpful metadata about your mapping run as [documented here](https://github.com/AllenInstitute/cell_type_mapper/blob/main/docs/output.md#json-output-file)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "86762dd5-e1b8-4dff-b356-72a2e1fb2a03",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(json_dst_path, 'rb') as src:\n",
    "    mapping_json = json.load(src)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dce8801-d6c5-4173-9179-55bd2e0c10cd",
   "metadata": {},
   "source": [
    "For instance, to see the configuration parameters corresponding to your mapping run, you can look at"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "9dfe4251-9411-4c4d-9c38-5d4bd0dee286",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"log_level\": \"ERROR\",\n",
      "  \"extended_result_path\": \"results/human_nonneurons_mapping.json\",\n",
      "  \"verbose_stdout\": true,\n",
      "  \"type_assignment\": {\n",
      "    \"bootstrap_factor\": 0.5,\n",
      "    \"log_level\": \"ERROR\",\n",
      "    \"bootstrap_factor_lookup\": null,\n",
      "    \"rng_seed\": 11235813,\n",
      "    \"chunk_size\": 10000,\n",
      "    \"bootstrap_iteration\": 100,\n",
      "    \"n_processors\": 4,\n",
      "    \"min_markers\": 10,\n",
      "    \"n_runners_up\": 5,\n",
      "    \"normalization\": \"raw\"\n",
      "  },\n",
      "  \"cloud_safe\": false,\n",
      "  \"hdf5_result_path\": null,\n",
      "  \"log_path\": null,\n",
      "  \"query_markers\": {\n",
      "    \"collapse_markers\": false,\n",
      "    \"log_level\": \"ERROR\",\n",
      "    \"serialized_lookup\": \"data/Human.query_markers.20250507.json\"\n",
      "  },\n",
      "  \"csv_result_path\": \"results/human_nonneruons_mapping.csv\",\n",
      "  \"extended_result_dir\": null,\n",
      "  \"query_gene_id_col\": null,\n",
      "  \"max_gb\": 100.0,\n",
      "  \"obsm_clobber\": false,\n",
      "  \"verbose_csv\": false,\n",
      "  \"drop_level\": null,\n",
      "  \"summary_metadata_path\": null,\n",
      "  \"map_to_ensembl\": false,\n",
      "  \"nodes_to_drop\": null,\n",
      "  \"precomputed_stats\": {\n",
      "    \"path\": \"data/Human.precomputed_stats.20250507.h5\",\n",
      "    \"log_level\": \"ERROR\"\n",
      "  },\n",
      "  \"tmp_dir\": null,\n",
      "  \"flatten\": false,\n",
      "  \"obsm_key\": null,\n",
      "  \"query_path\": \"/Users/scott.daniel/KnowledgeEngineering/HMBA_BasalGanglia_Consensus_Taxonomy/examples/abc_cache/expression_matrices/WHB-10Xv3/20240330/WHB-10Xv3-Nonneurons-raw.h5ad\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(mapping_json['config'], indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "417ab995-549b-4da5-b802-3a578c2e27d2",
   "metadata": {},
   "source": [
    "The actual cell type assignments are stored as a list under `'results'` as in"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c56ac2cf-c34f-4d80-ae9b-2a1c7dcf5cde",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"CCN20250428_LEVEL_0\": {\n",
      "    \"assignment\": \"CS20250428_NEIGH_0001\",\n",
      "    \"bootstrapping_probability\": 1.0,\n",
      "    \"aggregate_probability\": 1.0,\n",
      "    \"avg_correlation\": 0.7511638600643517,\n",
      "    \"runner_up_assignment\": [],\n",
      "    \"runner_up_correlation\": [],\n",
      "    \"runner_up_probability\": [],\n",
      "    \"directly_assigned\": true\n",
      "  },\n",
      "  \"CCN20250428_LEVEL_1\": {\n",
      "    \"assignment\": \"CS20250428_CLASS_0010\",\n",
      "    \"bootstrapping_probability\": 1.0,\n",
      "    \"aggregate_probability\": 1.0,\n",
      "    \"avg_correlation\": 0.8847336088608828,\n",
      "    \"runner_up_assignment\": [],\n",
      "    \"runner_up_correlation\": [],\n",
      "    \"runner_up_probability\": [],\n",
      "    \"directly_assigned\": true\n",
      "  },\n",
      "  \"CCN20250428_LEVEL_2\": {\n",
      "    \"assignment\": \"CS20250428_SUBCL_0022\",\n",
      "    \"bootstrapping_probability\": 1.0,\n",
      "    \"aggregate_probability\": 1.0,\n",
      "    \"avg_correlation\": 0.809104201147677,\n",
      "    \"runner_up_assignment\": [],\n",
      "    \"runner_up_correlation\": [],\n",
      "    \"runner_up_probability\": [],\n",
      "    \"directly_assigned\": true\n",
      "  },\n",
      "  \"CCN20250428_LEVEL_3\": {\n",
      "    \"assignment\": \"CS20250428_GROUP_0025\",\n",
      "    \"bootstrapping_probability\": 0.69,\n",
      "    \"aggregate_probability\": 0.69,\n",
      "    \"avg_correlation\": 0.626218144728874,\n",
      "    \"runner_up_assignment\": [\n",
      "      \"CS20250428_GROUP_0026\"\n",
      "    ],\n",
      "    \"runner_up_correlation\": [\n",
      "      0.6069362212706022\n",
      "    ],\n",
      "    \"runner_up_probability\": [\n",
      "      0.31\n",
      "    ],\n",
      "    \"directly_assigned\": true\n",
      "  },\n",
      "  \"CCN20250428_LEVEL_4\": {\n",
      "    \"assignment\": \"CS20250428_CLUST_0227\",\n",
      "    \"bootstrapping_probability\": 0.55,\n",
      "    \"aggregate_probability\": 0.3795,\n",
      "    \"avg_correlation\": 0.5614725591086703,\n",
      "    \"runner_up_assignment\": [\n",
      "      \"CS20250428_CLUST_0228\"\n",
      "    ],\n",
      "    \"runner_up_correlation\": [\n",
      "      0.5395855378047673\n",
      "    ],\n",
      "    \"runner_up_probability\": [\n",
      "      0.45\n",
      "    ],\n",
      "    \"directly_assigned\": true\n",
      "  },\n",
      "  \"cell_id\": \"10X362_3:TCAGTGAGTATTGACC\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(mapping_json['results'][0], indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa09ca1a-af74-4800-93c2-92dfdf0c646d",
   "metadata": {},
   "source": [
    "One complication is that the cell type assignments are only referred to by their unique machine-readable labels in this file. Fortunately, the cell type taxonomy, along with the mapping between machine- and human-readable cell type labels is also provided in this file. The `TaxonomyTree` class provides a helpful interface with that data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "12374288-0bbc-4fc7-ae96-1aa6e14b2eca",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cell_type_mapper.taxonomy.taxonomy_tree import TaxonomyTree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "85f6c1b9-6d8b-4c12-a3f5-075668b2afd5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/scott.daniel/miniconda3/envs/bg_documentation/lib/python3.12/site-packages/cell_type_mapper/taxonomy/utils.py:253: UserWarning: This taxonomy has no mapping from leaf_node -> rows in the cell by gene matrix\n",
      "  warnings.warn(\"This taxonomy has no mapping from leaf_node -> rows \"\n"
     ]
    }
   ],
   "source": [
    "taxonomy = TaxonomyTree(data=mapping_json['taxonomy_tree'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "beba2c6b-1945-44df-82ed-c4b85917b3d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Oligo OPALIN'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonomy.label_to_name(level='CCN20250428_LEVEL_3', label='CS20250428_GROUP_0025')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "108239a1-b373-4d1d-94fe-3928ea7326b7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
